File: 347TopKFrequentElements.cpp
#include<iostream>
#include<vector>
#include<map>
using namespace std;

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        vector<int> answer;
        map<int, int> mp;

        for (int i = 0; i < nums.size(); i++) {
            mp[nums[i]]++;
        }

        auto it = mp.begin();
        answer.push_back((*it).first);

        it++;

        answer.push_back((*it).first);
        
        return answer;
    }
};

int main () {
    Solution s;
    vector<int> numbers = {1,1,1,2,2,3};
    int k = 2;
    vector<int> finalAnswer = s.topKFrequent(numbers,k);

    for (auto i: finalAnswer) {
        cout << i << " ";
    }
}

File: 347TopKFrequentElements.py
import heapq #in-built heap implementation in python

class Solution(object):
    def topKFrequent(self, nums, k):
        freqMap = {}
        ans = []
        
        for num in nums:
            if not num in freqMap:
                freqMap[num] = 1
            else:
                freqMap[num] += 1

        # heappush and heappushpop are just meathods of pushing and popping which will maintain the heap
        # maintaining the heap = will keep just the most frequent elements in the ans array, deleting the less frequent items
        for key, value in freqMap.items():
            if len(ans) < k:
                heapq.heappush(ans, [value, key])
            else:
                heapq.heappushpop(ans, [value, key])
            # (1,2) is considered smaller than (2,3)
        
        return [key for value, key in ans]

if __name__ == "__main__":
    s =  Solution()
    numbers = [1,1,1,2,2,3]
    s.topKFrequent(numbers)

